File: sdks/python/src/touchenv/crypto.py
"""AES-256-GCM encryption and decryption for touchenv."""

from __future__ import annotations

import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

NONCE_LENGTH = 12
TAG_LENGTH = 16
KEY_LENGTH = 32


def parse_key(hex_key: str) -> bytes:
    """Parse a hex-encoded 256-bit DEK into bytes.

    Raises ValueError if the key is not exactly 64 hex characters.
    """
    hex_key = hex_key.strip()
    if len(hex_key) != 64:
        raise ValueError("DEK must be exactly 64 hex characters (256 bits)")
    try:
        return bytes.fromhex(hex_key)
    except ValueError as exc:
        raise ValueError("DEK must be exactly 64 hex characters (256 bits)") from exc


def encrypt(
    key: bytes,
    plaintext: bytes,
    aad: bytes,
    nonce: bytes | None = None,
) -> tuple[bytes, bytes, bytes]:
    """Encrypt plaintext using AES-256-GCM.

    Returns (nonce, ciphertext, tag) where ciphertext does NOT include the tag
    (tag is separated from the AESGCM output which appends it).
    """
    if len(key) != KEY_LENGTH:
        raise ValueError(f"Key must be {KEY_LENGTH} bytes")
    iv = nonce if nonce is not None else os.urandom(NONCE_LENGTH)
    if len(iv) != NONCE_LENGTH:
        raise ValueError(f"Nonce must be {NONCE_LENGTH} bytes")

    aesgcm = AESGCM(key)
    # AESGCM.encrypt returns ciphertext || tag (tag is last 16 bytes)
    ct_with_tag = aesgcm.encrypt(iv, plaintext, aad)
    ciphertext = ct_with_tag[:-TAG_LENGTH]
    tag = ct_with_tag[-TAG_LENGTH:]
    return iv, ciphertext, tag


def decrypt(
    key: bytes,
    nonce: bytes,
    ciphertext: bytes,
    tag: bytes,
    aad: bytes,
) -> bytes:
    """Decrypt ciphertext using AES-256-GCM.

    Returns the decrypted plaintext bytes.
    Raises an exception if authentication fails.
    """
    if len(key) != KEY_LENGTH:
        raise ValueError(f"Key must be {KEY_LENGTH} bytes")
    if len(nonce) != NONCE_LENGTH:
        raise ValueError(f"Nonce must be {NONCE_LENGTH} bytes")
    if len(tag) != TAG_LENGTH:
        raise ValueError(f"Tag must be {TAG_LENGTH} bytes")

    aesgcm = AESGCM(key)
    # AESGCM.decrypt expects ciphertext || tag
    ct_with_tag = ciphertext + tag
    return aesgcm.decrypt(nonce, ct_with_tag, aad)

File: cli/src/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes } from 'node:crypto';

const ALGORITHM = 'aes-256-gcm';
const NONCE_LENGTH = 12;
const TAG_LENGTH = 16;
const KEY_LENGTH = 32;

export interface EncryptResult {
  nonce: Buffer;
  ciphertext: Buffer;
  tag: Buffer;
}

/**
 * Parse a hex-encoded 256-bit DEK into a Buffer.
 * Throws if the key is not exactly 64 hex characters.
 */
export function parseKey(hex: string): Buffer {
  if (!/^[0-9a-fA-F]{64}$/.test(hex)) {
    throw new Error('DEK must be exactly 64 hex characters (256 bits)');
  }
  return Buffer.from(hex, 'hex');
}

/**
 * Encrypt plaintext using AES-256-GCM.
 *
 * @param key    - 32-byte DEK
 * @param plain  - UTF-8 plaintext buffer
 * @param aad    - Additional Authenticated Data (header bytes)
 * @param nonce  - Optional 12-byte nonce (random if omitted)
 */
export function encrypt(
  key: Buffer,
  plain: Buffer,
  aad: Buffer,
  nonce?: Buffer,
): EncryptResult {
  if (key.length !== KEY_LENGTH) {
    throw new Error(`Key must be ${KEY_LENGTH} bytes`);
  }
  const iv = nonce ?? randomBytes(NONCE_LENGTH);
  if (iv.length !== NONCE_LENGTH) {
    throw new Error(`Nonce must be ${NONCE_LENGTH} bytes`);
  }

  const cipher = createCipheriv(ALGORITHM, key, iv, { authTagLength: TAG_LENGTH });
  cipher.setAAD(aad);

  const encrypted = Buffer.concat([cipher.update(plain), cipher.final()]);
  const tag = cipher.getAuthTag();

  return { nonce: iv, ciphertext: encrypted, tag };
}

/**
 * Decrypt ciphertext using AES-256-GCM.
 *
 * @param key        - 32-byte DEK
 * @param nonce      - 12-byte IV
 * @param ciphertext - encrypted payload
 * @param tag        - 16-byte GCM auth tag
 * @param aad        - Additional Authenticated Data (header bytes)
 * @returns Decrypted plaintext buffer
 */
export function decrypt(
  key: Buffer,
  nonce: Buffer,
  ciphertext: Buffer,
  tag: Buffer,
  aad: Buffer,
): Buffer {
  if (key.length !== KEY_LENGTH) {
    throw new Error(`Key must be ${KEY_LENGTH} bytes`);
  }
  if (nonce.length !== NONCE_LENGTH) {
    throw new Error(`Nonce must be ${NONCE_LENGTH} bytes`);
  }
  if (tag.length !== TAG_LENGTH) {
    throw new Error(`Tag must be ${TAG_LENGTH} bytes`);
  }

  const decipher = createDecipheriv(ALGORITHM, key, nonce, { authTagLength: TAG_LENGTH });
  decipher.setAAD(aad);
  decipher.setAuthTag(tag);

  const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  return decrypted;
}

export function generateKey(): Buffer {
  return randomBytes(KEY_LENGTH);
}

export function generateNonce(): Buffer {
  return randomBytes(NONCE_LENGTH);
}
